src/main/java/io/strimzi/kproxy/internal/interceptor/ProduceRecordTransformationInterceptor.java

```diff
@@ -64,38 +64,37 @@ public boolean shouldDecodeResponse(ApiKeys apiKey, int apiVersion) {
     }
 
     @Override
-    public ChannelInboundHandler frontendHandler() {
-        return new ChannelInboundHandlerAdapter() {
+    public RequestHandler requestHandler() {
+        return new RequestHandler() {
+
             @Override
-            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
-                if (msg instanceof DecodedRequestFrame && ((DecodedRequestFrame) msg).apiKey() == ApiKeys.PRODUCE) {
-                    var req = (ProduceRequestData) ((DecodedRequestFrame) msg).body();
-
-                    req.topicData().forEach(tpd -> {
-                        for (PartitionProduceData partitionData : tpd.partitionData()) {
-                            MemoryRecords records = (MemoryRecords) partitionData.records();
-                            MemoryRecordsBuilder newRecords = MemoryRecords.builder(ByteBuffer.allocate(records.sizeInBytes()), CompressionType.NONE,
-                                    TimestampType.CREATE_TIME, 0);
-
-                            for (MutableRecordBatch batch : records.batches()) {
-                                for (Iterator<Record> batchRecords = batch.iterator(); batchRecords.hasNext();) {
-                                    Record batchRecord = batchRecords.next();
-                                    newRecords.append(batchRecord.timestamp(), batchRecord.key(), valueTransformation.transformation(batchRecord.value()));
-                                }
-                            }
+            public DecodedRequestFrame handleRequest(DecodedRequestFrame requestFrame) {
+                var req = (ProduceRequestData) requestFrame.body();
+
+                req.topicData().forEach(tpd -> {
+                    for (PartitionProduceData partitionData : tpd.partitionData()) {
+                        MemoryRecords records = (MemoryRecords) partitionData.records();
+                        MemoryRecordsBuilder newRecords = MemoryRecords.builder(ByteBuffer.allocate(records.sizeInBytes()), CompressionType.NONE,
+                                TimestampType.CREATE_TIME, 0);
```

Comment:
Consider https://github.com/a0x8o/kafka/blob/944329fb01c2666cc9d9646a98879b985a3e97c0/clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java#L56-L58 and if makes sense to use the channel's allocator to allocate a single chunk here (or just trying to eagerly allocate a right sized buffer)

Reply:
Interesting, not sure though whether I follow. It is using `ByteBuffer.allocate(records.sizeInBytes())` above, which should allocate the right buffer size, if that transformation doesn't expand the size of the actual message payload (as is the case for the simple upper-casing example). The thing is, here we don't quite know whether that is the case or not, given a `Function` for executing the transformation is passed to this class. Or are you saying we should pass maybe another `Function` for obtaining the buffer size to the constructor of `ProduceRecordTransformationInterceptor`?

Reply:
@gunnarmorling I _think_ the point @franz1981 is making is that here we allocate a new `ByteBuffer` on the heap every time. A netty channel has an allocator that is potentially more efficient (e.g. using a pool of buffers and/or allocating off heap, thus less gc). The problem is the mismatch between the `ByteBuf` that netty allocates vs the `ByteBuffer` that `MemoryRecords` requires. Is that right @franz1981?

Reply:
@franz1981, ok, so now I'm obtaining the `ByteBuffer` via Netty. Is that what you had in mind?